replace secret copying with loading from file for ImagePullSecret

mrIncompetent opened the pull request:
**What this PR does / why we need it**:
Changes the way we handle ImagePullSecrets for user-clusters.
In the old way, we copied a Secret from the Kubermatic namespace.
Though this worked, it required the initial ImagePullSecret to exist already inside `kubermatic` namespace within the cluster.

This approach will load the ImagePullSecret from a file, specified via a flag.
The Kubermatic Helm chart got modified to mount the already existing ImagePullSecret to the kubermatic-controller-manager.

Also this PR will combine the existing ImagePullSecrets for Quay.io & Docker-Hub.

If this PR got approved, a second PR for the Secrets repository will be opened.

**Release note**:
```release-note
Combine ImagePullSecrets im the Kubermatic chart.
```

mrIncompetent committed:
replace secret copying with loading from file for ImagePullSecret

alvaroaleman commented on api/cmd/kubermatic-controller-manager/main.go:
```diff
@@ -61,6 +61,7 @@ type controllerRunOptions struct {
 	etcdDiskSize                           string
 	inClusterPrometheusRulesFile           string
 	inClusterPrometheusDisableDefaultRules bool
+	imagePullSecretFile                    string
```
Ugh. What is the problem with creating it from the lister? An `imagePullSecretName` flag is obvious in its meaning for everyone since there is even a secret type for it but `imagePullSecretFile`? Is it the raw docker json? Is it a a serialized kube secret on disk? This change feels like we move complexity from the code into the deployment

mrIncompetent replied:
> What is the problem with creating it from the lister?

The problem with creating it from the lister:
- it introduces a new source for resources - objects from the kubermatic namespace
- It requires that specific resources need to magically exist in the kubermatic namespace
  - Either the kubermatic chat needs to be deployed
  - Or the namespace needs to exist with the secret in it
- It creates a hard binding between the Deployment & the application itself (By the dependency on the objects in the namespace)


In Kubernetes file dependencies for an application should be mounted into the pod.
The application should not talk to Kubernetes and load it via API.


That all above being said, in this scenario it does not hurt.
But i want to draw that line very early and clearly.


> `imagePullSecretFile`? Is it the raw docker json? Is it a a serialized kube secret on disk? 

Its the raw docker json. Maybe `dockerConfigJsonFile` would be a better name?

alvaroaleman replied:
> it introduces a new source for resources - objects from the kubermatic namespace

That's the only valid point IMHO because for the others

> It requires that specific resources need to magically exist in the kubermatic namespace
> It creates a hard binding between the Deployment & the application itself (By the dependency on the objects in the namespace)

They are now not required to "magically exist in the kubermatic namespace" but they are required to "magically exist on disk". And the name/namespace of the secret can be made configurable the same way you made the filepath configurable.

Generally speaking both are valid but IMHO the lister approach has the advantage that we can update the secret without restarting the controller which is why I prefer it as source.

> Its the raw docker json. Maybe `dockerConfigJsonFile` would be a better name?

What about `dockerPullConfigJsonFile`?

mrIncompetent committed:
rename image-pull-secret to docker-pull-config-json-file

mrIncompetent committed:
roll controller-manager deployment when the dockercfg secret changes

mrIncompetent commented:
Renamed flag to `docker-pull-config-json-file` and the internal variable to dockerPullConfigJSONFile in https://github.com/kubermatic/kubermatic/pull/1877/commits/2300974c4c391e05477a603ecd05810980c74d0f.

Added the checksum to the controller-manager deployment in https://github.com/kubermatic/kubermatic/pull/1877/commits/1e2dd27ce4ba243b5006a4d96bc2faf7f68a3984

PTAL

alvaroaleman reviewed (approved):
Well I still think using data from the Kube api is the superior approach as it means it can be updated without redeploying but I wont block on that.
